# Foundations of Computer Science Project
### Kiva Dataset
Paolo Mariani
<br>Matricola 800307

#### Import all required libraries

In [122]:
import pandas as pd
import re
import datetime as dt

#### Import all required Datasets

In [123]:
loan_lenders = pd.read_csv("kiva/loans_lenders.csv")
loan_lenders.head()

loan_id                                            lenders
0   483693  muc888, sam4326, camaran3922, lachheb1865, reb...
1   483738  muc888, nora3555, williammanashi, barbara5610,...
2   485000  muc888, terrystl, richardandsusan8352, sherri4...
3   486087  muc888, james5068, rudi5955, daniel9859, don92...
4   534428  muc888, niki3008, teresa9174, mike4896, david7...

In [124]:
loans = pd.read_csv("kiva/loans.csv")
loans.head()

loan_id       loan_name original_language  \
0   657307            Aivy           English   
1   657259  Idalia Marizza           Spanish   
2   658010           Aasia           English   
3   659347         Gulmira           Russian   
4   656933         Ricky\t           English   

                                         description  \
0  Aivy, 21 years of age, is single and lives in ...   
1  Doña Idalia, esta casada, tiene 57 años de eda...   
2  Aasia is a 45-year-old married lady and she ha...   
3  Гулмире 36 лет, замужем, вместе с супругом вос...   
4  Ricky is a farmer who currently cultivates his...   

                              description_translated  funded_amount  \
0                                                NaN          125.0   
1  Idalia, 57, is married and lives with her husb...          400.0   
2                                                NaN          400.0   
3  Gulmira is 36 years old and married.  She and ...          625.0   
4                                                NaN          425.0   

   loan_amount  status  activity_name  sector_name  ...  \
0        125.0  funded  General Store       Retail  ...   
1        400.0  funded  Used Clothing     Clothing  ...   
2        400.0  funded  General Store       Retail  ...   
3        625.0  funded        Farming  Agriculture  ...   
4        425.0  funded        Farming  Agriculture  ...   

                     raised_time lender_term num_lenders_total  \
0  2014-01-15 04:48:22.000 +0000         7.0                 3   
1  2014-02-25 06:42:06.000 +0000         8.0                11   
2  2014-01-24 23:06:18.000 +0000        14.0                16   
3  2014-01-22 05:29:28.000 +0000        14.0                21   
4  2014-01-14 17:29:27.000 +0000         7.0                15   

  num_journal_entries num_bulk_entries  \
0                   2                1   
1                   2                1   
2                   2                1   
3                   2                1   
4                   2                1   

                                                tags borrower_genders  \
0                                                NaN           female   
1                                                NaN           female   
2  #Woman Owned Biz, #Supporting Family, user_fav...           female   
3                                      user_favorite           female   
4           #Animals, #Eco-friendly, #Sustainable Ag             male   

   borrower_pictured repayment_interval distribution_model  
0               true          irregular      field_partner  
1               true            monthly      field_partner  
2               true            monthly      field_partner  
3               true            monthly      field_partner  
4               true             bullet      field_partner  

[5 rows x 31 columns]

In [125]:
lenders = pd.read_csv("kiva/lenders.csv")
lenders.head()

permanent_name display_name city state country_code  member_since  \
0        qian3013         Qian  NaN   NaN          NaN    1461300457   
1       reena6733        Reena  NaN   NaN          NaN    1461300634   
2         mai5982          Mai  NaN   NaN          NaN    1461300853   
3  andrew86079135       Andrew  NaN   NaN          NaN    1461301091   
4      nguyen6962       Nguyen  NaN   NaN          NaN    1461301154   

  occupation loan_because  loan_purchase_num invited_by  num_invited  
0        NaN          NaN                1.0        NaN            0  
1        NaN          NaN                9.0        NaN            0  
2        NaN          NaN                NaN        NaN            0  
3        NaN          NaN                5.0  Peter Tan            0  
4        NaN          NaN                NaN        NaN            0

### 1) Normalize the loan_lenders table. In the normalized table, each row must have one loan_id and one lender

The Dataset must be normalized, in each value of the column "lenders" there's a list of users which has to be normalized and split in different rows (one for each user).

In [126]:
lenders_list = loan_lenders["lenders"].str.split(",").explode()
lenders_list.head()

0          muc888
0         sam4326
0     camaran3922
0     lachheb1865
0     rebecca3499
Name: lenders, dtype: object

In [138]:
loan_lenders_norm = pd.merge(loan_lenders, lenders_list, left_index=True, right_index=True)
loan_lenders_norm

loan_id                                          lenders_x  \
0         483693  muc888, sam4326, camaran3922, lachheb1865, reb...   
0         483693  muc888, sam4326, camaran3922, lachheb1865, reb...   
0         483693  muc888, sam4326, camaran3922, lachheb1865, reb...   
0         483693  muc888, sam4326, camaran3922, lachheb1865, reb...   
0         483693  muc888, sam4326, camaran3922, lachheb1865, reb...   
...          ...                                                ...   
1387430  1206425  rich6705, sergiiy9766, angela7509, barbara5610...   
1387430  1206425  rich6705, sergiiy9766, angela7509, barbara5610...   
1387430  1206425  rich6705, sergiiy9766, angela7509, barbara5610...   
1387431  1206486                              alan5175, amy38101311   
1387431  1206486                              alan5175, amy38101311   

                  lenders_y  
0                    muc888  
0                   sam4326  
0               camaran3922  
0               lachheb1865  
0               rebecca3499  
...                     ...  
1387430   trogdorfamily7622  
1387430           danny6470  
1387430             don6118  
1387431            alan5175  
1387431         amy38101311  

[28293931 rows x 3 columns]

To verify the normalization i do print the size of the list for the first row (index 0) and after i print the number of rows that are obtained with index 0.

In [139]:
len(loan_lenders.loc[0, "lenders"].split(","))

40

In [140]:
loan_lenders_norm.groupby(loan_lenders_norm.index).count().head()

loan_id  lenders_x  lenders_y
0       40         40         40
1       15         15         15
2       15         15         15
3       13         13         13
4       19         19         19

Then the column "lenders_x" is deleted and "lenders_y" becomes "lender":

In [141]:
loan_lenders_norm.drop(["lenders_x"], axis = 1, inplace = True)
loan_lenders_norm.rename({"lenders_y": "lender"}, axis = 1, inplace = True)
loan_lenders_norm.reset_index(inplace = True)
loan_lenders_norm.drop(["index"], axis = 1, inplace = True)
loan_lenders_norm.head()

loan_id        lender
0   483693        muc888
1   483693       sam4326
2   483693   camaran3922
3   483693   lachheb1865
4   483693   rebecca3499

### 2) For each loan, add a column duration corresponding to the number of days between the disburse time and the planned expiration time. If any of those two dates is missing, also the duration must be missing

In [143]:
loans.head()[["loan_id", "loan_name", "planned_expiration_time", "disburse_time"]]

loan_id       loan_name        planned_expiration_time  \
0   657307            Aivy  2014-02-14 03:30:06.000 +0000   
1   657259  Idalia Marizza  2014-03-26 22:25:07.000 +0000   
2   658010           Aasia  2014-02-15 21:10:05.000 +0000   
3   659347         Gulmira  2014-02-21 03:10:02.000 +0000   
4   656933         Ricky\t  2014-02-13 06:10:02.000 +0000   

                   disburse_time  
0  2013-12-22 08:00:00.000 +0000  
1  2013-12-20 08:00:00.000 +0000  
2  2014-01-09 08:00:00.000 +0000  
3  2014-01-17 08:00:00.000 +0000  
4  2013-12-17 08:00:00.000 +0000

The duration attribute is computed with a simple difference between "planned_expiration_time" and "disburse_time" in datetime format. All the values of these attributes that are missing automatically set the "duration" attribute as None.

In [144]:
loans["planned_expiration_time"] = pd.to_datetime(loans["planned_expiration_time"])
loans["disburse_time"] = pd.to_datetime(loans["disburse_time"])

In [145]:
loans["duration"] = loans["planned_expiration_time"] - loans["disburse_time"]
loans[loans["planned_expiration_time"].isnull()]["duration"] == None
loans[loans["disburse_time"].isnull()]["duration"] == None
print(len(loans[loans.duration.isnull()]))
loans.duration.head()

374645


0   53 days 19:30:06
1   96 days 14:25:07
2   37 days 13:10:05
3   34 days 19:10:02
4   57 days 22:10:02
Name: duration, dtype: timedelta64[ns]

In [146]:
len(loans[loans.duration.isnull()])/len(loans)

0.26390754624343216

There are currently 374645 missing values (26.39% of the dataset). I do compute the day attribute from the "duration" column by obtaining the .days attribute with a lambda function:

In [147]:
loans["duration"] = loans.apply(lambda x: x["duration"].days, axis = 1)
loans["duration"]

0          53.0
1          96.0
2          37.0
3          34.0
4          57.0
           ... 
1419602    39.0
1419603    39.0
1419604    51.0
1419605    63.0
1419606    61.0
Name: duration, Length: 1419607, dtype: float64

In [148]:
len(loans[loans.planned_expiration_time<loans.disburse_time])

14935

There are also 14935 values where the duration time is negative

The result of the computing is the following:

In [150]:
loans.head()[["loan_id", "loan_name", "planned_expiration_time", "disburse_time", "duration"]]

loan_id       loan_name   planned_expiration_time  \
0   657307            Aivy 2014-02-14 03:30:06+00:00   
1   657259  Idalia Marizza 2014-03-26 22:25:07+00:00   
2   658010           Aasia 2014-02-15 21:10:05+00:00   
3   659347         Gulmira 2014-02-21 03:10:02+00:00   
4   656933         Ricky\t 2014-02-13 06:10:02+00:00   

              disburse_time  duration  
0 2013-12-22 08:00:00+00:00      53.0  
1 2013-12-20 08:00:00+00:00      96.0  
2 2014-01-09 08:00:00+00:00      37.0  
3 2014-01-17 08:00:00+00:00      34.0  
4 2013-12-17 08:00:00+00:00      57.0

### 3) Find the lenders that have funded at least twice

In [151]:
loan_lenders_norm.head()

loan_id        lender
0   483693        muc888
1   483693       sam4326
2   483693   camaran3922
3   483693   lachheb1865
4   483693   rebecca3499

I do calculate the number of loans made by each lender by grouping them by the lender's name, then counting the number of items in each group:

In [152]:
lenders_funds = loan_lenders_norm.groupby("lender")["loan_id"].count().reset_index()
lenders_funds.columns = ["lender", "fund_count"]
lenders_funds.head()

lender  fund_count
0              000          39
1            00000          39
2             0002          70
3         00mike00           1
4   0101craign0101          71

Then, the result is filtered to obtain only the lenders which have made more than one loan:

In [153]:
lenders_funds[lenders_funds["fund_count"]>1][["lender", "fund_count"]].head()

lender  fund_count
0              000          39
1            00000          39
2             0002          70
4   0101craign0101          71
5          0132575           4

### 4) For each country, compute how many loans have involved that country as borrowers

In [159]:
print(f"The number of loans that have unknown country name is {loans.country_name.isnull().sum()}")
loans.country_name.value_counts().reset_index()

The number of loans that have unknown country name is 0


index  country_name
0        Philippines        285336
1              Kenya        143699
2               Peru         86000
3           Cambodia         79701
4        El Salvador         64037
..               ...           ...
91  Papua New Guinea             1
92           Uruguay             1
93          Botswana             1
94        Mauritania             1
95            Canada             1

[96 rows x 2 columns]

There are no missing values in the "country_name" column, most of the loans are distributed in Philippines, Kenya and Perù. <br> To compute how many times each country is involved in loans as borrower the "loans" dataset is grouped by the attribute "country_name":

In [165]:
country_count = loans.groupby("country_name")["loan_id"].count().reset_index()
country_count.columns = ["country_name", "count"]
country_count.head()

country_name  count
0  Afghanistan   2337
1      Albania   3075
2      Armenia  13952
3   Azerbaijan  10172
4       Belize    218

### 5) For each country, compute the overall amount of money borrowed

To compute the result, the "loans" dataset must be grouped by the "country_name" column. Here is proposed a simplified version of the "loans" dataframe with the interesting features:

In [173]:
loans[["loan_id", "country_name", "disburse_time", "planned_expiration_time", "funded_amount", "loan_amount", "currency", "duration"]]

loan_id country_name             disburse_time  \
0         657307  Philippines 2013-12-22 08:00:00+00:00   
1         657259     Honduras 2013-12-20 08:00:00+00:00   
2         658010     Pakistan 2014-01-09 08:00:00+00:00   
3         659347   Kyrgyzstan 2014-01-17 08:00:00+00:00   
4         656933  Philippines 2013-12-17 08:00:00+00:00   
...          ...          ...                       ...   
1419602   988180        Kenya 2015-11-23 08:00:00+00:00   
1419603   988213  Philippines 2015-11-24 08:00:00+00:00   
1419604   989109        Ghana 2015-11-13 08:00:00+00:00   
1419605   989143  Philippines 2015-11-03 08:00:00+00:00   
1419606   989240   Madagascar 2015-11-03 08:00:00+00:00   

          planned_expiration_time  funded_amount  loan_amount currency  \
0       2014-02-14 03:30:06+00:00          125.0        125.0      PHP   
1       2014-03-26 22:25:07+00:00          400.0        400.0      HNL   
2       2014-02-15 21:10:05+00:00          400.0        400.0      PKR   
3       2014-02-21 03:10:02+00:00          625.0        625.0      KGS   
4       2014-02-13 06:10:02+00:00          425.0        425.0      PHP   
...                           ...            ...          ...      ...   
1419602 2016-01-02 01:00:03+00:00          400.0        400.0      KES   
1419603 2016-01-02 16:40:07+00:00          300.0        300.0      PHP   
1419604 2016-01-03 22:20:04+00:00         2425.0       2425.0      GHS   
1419605 2016-01-05 08:50:02+00:00          100.0        100.0      PHP   
1419606 2016-01-03 20:50:06+00:00          175.0        175.0      MGA   

         duration  
0            53.0  
1            96.0  
2            37.0  
3            34.0  
4            57.0  
...           ...  
1419602      39.0  
1419603      39.0  
1419604      51.0  
1419605      63.0  
1419606      61.0  

[1419607 rows x 8 columns]

It is necessary to consider if items from the dataset contain null values for "loan_amount":

In [186]:
print(f"There are {loans.loan_amount.isnull().sum()} null values in the loans dataframe")

There are 0 null values in the loans dataframe


The result is computed with a group_by and a sum function evaluated on it:

In [196]:
loan_per_country = loans.groupby("country_name")["country_name", "funded_amount", "loan_amount"].sum().reset_index()
loan_per_country.head()

C:\Users\paolo\anaconda3\lib\site-packages\ipykernel_launcher.py:1: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  """Entry point for launching an IPython kernel.


country_name  funded_amount  loan_amount
0  Afghanistan      1964650.0    1967950.0
1      Albania      3992475.0    4307350.0
2      Armenia     20284675.0   22950475.0
3   Azerbaijan     14191350.0   14784625.0
4       Belize       150175.0     150175.0

### 6) Like the previous point, but expressed as a percentage of the overall amount lent

In [247]:
total_loans_amount = loan_per_country["loan_amount"].sum()
total_funded_amount = loan_per_country["funded_amount"].sum()

In [248]:
loan_per_country["funded_amount_perc"] = round(loan_per_country["funded_amount"]/total_funded_amount*100, 3)
loan_per_country["loan_amount_perc"] = round(loan_per_country["loan_amount"]/total_loans_amount*100, 3)
loan_per_country.head()

country_name  funded_amount  loan_amount  funded_amount_perc  \
0  Afghanistan      1964650.0    1967950.0               0.174   
1      Albania      3992475.0    4307350.0               0.353   
2      Armenia     20284675.0   22950475.0               1.795   
3   Azerbaijan     14191350.0   14784625.0               1.256   
4       Belize       150175.0     150175.0               0.013   

   loan_amount_perc  
0             0.167  
1             0.365  
2             1.943  
3             1.251  
4             0.013

In [249]:
loan_per_country.sort_values(by = "loan_amount_perc", ascending=False)

country_name  funded_amount  loan_amount  funded_amount_perc  \
66     Philippines     97045925.0   97984600.0               8.587   
65            Peru     78702800.0   79437775.0               6.964   
40           Kenya     63650255.0   66735975.0               5.632   
64        Paraguay     53715200.0   53964700.0               4.753   
14        Cambodia     50829425.0   51613525.0               4.497   
..             ...            ...          ...                 ...   
89         Uruguay         8000.0       8000.0               0.001   
90         Vanuatu         9250.0       9250.0               0.001   
92  Virgin Islands            0.0      10000.0               0.000   
31            Guam         8700.0      17300.0               0.001   
28            Gaza         5000.0       5000.0               0.000   

    loan_amount_perc  
66             8.294  
65             6.724  
40             5.649  
64             4.568  
14             4.369  
..               ...  
89             0.001  
90             0.001  
92             0.001  
31             0.001  
28             0.000  

[96 rows x 5 columns]

As can be seen, most of the loans are made in Philippines, Peru and Kenya (the top 3 contries by loans requested)

### 7) Like the three previous points, but split for each year (with respect to disburse time)

In [240]:
print(f"There are {loans.disburse_time.isnull().sum()} null values of \"disbursed_time\", so these items will not be considered to compute the result")

There are 2813 null values of "disbursed_time", so these items will not be considered to compute the result


The rows with null values in "disburse_time" are dropped, another dataset ("loans_nonull") is created for this purpose to avoid losing these informations:

In [243]:
loans_nonull = loans[loans.disburse_time.notnull()]
len(loans_nonull)

1416794

The result can be obtained by grouping the "loans" dataframe by "year";
<br> In the first case how many loans have involved that country as borrowers is computed:

In [253]:
loans_by_year = loans_nonull.groupby(loans.disburse_time.dt.year)["loan_id"].count().reset_index()
loans_by_year.columns = ["year", "number_of_loans"]
loans_by_year.year = loans_by_year.year.astype(int)
loans_by_year

year  number_of_loans
0   2005              203
1   2006             2172
2   2007            24400
3   2008            54586
4   2009            83076
5   2010            93466
6   2011           114540
7   2012           133650
8   2013           140167
9   2014           172709
10  2015           183671
11  2016           197187
12  2017           216003
13  2018              964

For the second task, the overall amount of money borrowed for each year is computed:

In [254]:
loans_by_year["funded_amount"] = loans_nonull.groupby(loans_nonull.disburse_time.dt.year)["funded_amount"].sum().reset_index()["funded_amount"]
loans_by_year["loan_amount"] = loans_nonull.groupby(loans_nonull.disburse_time.dt.year)["loan_amount"].sum().reset_index()["loan_amount"]
loans_by_year.head()

year  number_of_loans  funded_amount  loan_amount
0  2005              203       102850.0     102850.0
1  2006             2172      1375175.0    1376575.0
2  2007            24400     15439425.0   15446525.0
3  2008            54586     39384425.0   39423050.0
4  2009            83076     59477925.0   59689475.0

For the third task the total amount of loan and funded amounts is computed; then the loan_amount and funded_amount values for each year is divided by the respective total amount

In [255]:
tot_funded = loans_by_year["funded_amount"].sum()
tot_loan = loans_by_year["loan_amount"].sum()
print(tot_funded, tot_loan)

1128191775.0 1170496550.0


In [256]:
loans_by_year["funded_amount_perc"] = round(loans_by_year["funded_amount"]/tot_funded*100, 3)
loans_by_year["loan_amount_perc"] = round(loans_by_year["loan_amount"]/tot_loan*100, 3)
loans_by_year

year  number_of_loans  funded_amount  loan_amount  funded_amount_perc  \
0   2005              203       102850.0     102850.0               0.009   
1   2006             2172      1375175.0    1376575.0               0.122   
2   2007            24400     15439425.0   15446525.0               1.369   
3   2008            54586     39384425.0   39423050.0               3.491   
4   2009            83076     59477925.0   59689475.0               5.272   
5   2010            93466     72187300.0   72609150.0               6.398   
6   2011           114540     93601250.0   93699300.0               8.297   
7   2012           133650    115427450.0  119977575.0              10.231   
8   2013           140167    128301200.0  132043925.0              11.372   
9   2014           172709    146668375.0  152270425.0              13.000   
10  2015           183671    148808700.0  155835650.0              13.190   
11  2016           197187    151001875.0  160978375.0              13.384   
12  2017           216003    155644275.0  166053650.0              13.796   
13  2018              964       771550.0     990025.0               0.068   

    loan_amount_perc  
0              0.009  
1              0.118  
2              1.320  
3              3.368  
4              5.100  
5              6.203  
6              8.005  
7             10.250  
8             11.281  
9             13.009  
10            13.314  
11            13.753  
12            14.187  
13             0.085

### 8) For each lender, compute the overall amount of money lent. For each loan that has more than one lender, you must assume that all lenders contributed the same amount

In [267]:
loans.head()

loan_id       loan_name original_language  \
0   657307            Aivy           English   
1   657259  Idalia Marizza           Spanish   
2   658010           Aasia           English   
3   659347         Gulmira           Russian   
4   656933         Ricky\t           English   

                                         description  \
0  Aivy, 21 years of age, is single and lives in ...   
1  Doña Idalia, esta casada, tiene 57 años de eda...   
2  Aasia is a 45-year-old married lady and she ha...   
3  Гулмире 36 лет, замужем, вместе с супругом вос...   
4  Ricky is a farmer who currently cultivates his...   

                              description_translated  funded_amount  \
0                                                NaN          125.0   
1  Idalia, 57, is married and lives with her husb...          400.0   
2                                                NaN          400.0   
3  Gulmira is 36 years old and married.  She and ...          625.0   
4                                                NaN          425.0   

   loan_amount  status  activity_name  sector_name  ... lender_term  \
0        125.0  funded  General Store       Retail  ...         7.0   
1        400.0  funded  Used Clothing     Clothing  ...         8.0   
2        400.0  funded  General Store       Retail  ...        14.0   
3        625.0  funded        Farming  Agriculture  ...        14.0   
4        425.0  funded        Farming  Agriculture  ...         7.0   

  num_lenders_total num_journal_entries num_bulk_entries  \
0                 3                   2                1   
1                11                   2                1   
2                16                   2                1   
3                21                   2                1   
4                15                   2                1   

                                                tags  borrower_genders  \
0                                                NaN            female   
1                                                NaN            female   
2  #Woman Owned Biz, #Supporting Family, user_fav...            female   
3                                      user_favorite            female   
4           #Animals, #Eco-friendly, #Sustainable Ag              male   

  borrower_pictured  repayment_interval distribution_model duration  
0              true           irregular      field_partner     53.0  
1              true             monthly      field_partner     96.0  
2              true             monthly      field_partner     37.0  
3              true             monthly      field_partner     34.0  
4              true              bullet      field_partner     57.0  

[5 rows x 32 columns]

The number of lenders for each loan si calculated with a simple group_by on the loan_id:

In [278]:
loans_lender_number = loan_lenders_norm.groupby("loan_id").count().reset_index()
loans_lender_number.head()

loan_id  lender
0       84       3
1       85       2
2       86       3
3       88       3
4       89       4

In [279]:
print(f"There are {len(loans_lender_number)} different loans made in the database")

There are 1387432 different loans made in the database


In [284]:
loans_aug = pd.merge(loans_lender_number, loans, on = "loan_id")[["loan_id", "lender", "loan_name", "funded_amount", "loan_amount", "num_lenders_total", "duration"]]
loans_aug

loan_id  lender loan_name  funded_amount  loan_amount  \
0             84       3   Justine          500.0        500.0   
1             85       2  Geoffrey          500.0        500.0   
2             86       3    Apollo          500.0        500.0   
3             88       3    Eunice          300.0        300.0   
4             89       4      Rose          500.0        500.0   
...          ...     ...       ...            ...          ...   
1387423  1444051       1    QAViva           50.0        225.0   
1387424  1444053       1    QAViva           50.0        725.0   
1387425  1444058       1       NaN           25.0         25.0   
1387426  1444063       1    QAViva           50.0         50.0   
1387427  1444065       1       NaN           25.0         25.0   

         num_lenders_total  duration  
0                        5       NaN  
1                        3       NaN  
2                        3       NaN  
3                        4       NaN  
4                        6       NaN  
...                    ...       ...  
1387423                  2      30.0  
1387424                  2      30.0  
1387425                  1      30.0  
1387426                  2      30.0  
1387427                  1      30.0  

[1387428 rows x 7 columns]

There are multiple cases in which the number of lenders obtained in "loans_lender_number" is different by the "num_lenders_total" attribute:

In [289]:
print(f"The number of different lenders number is {len(loans_aug[loans_aug.lender != loans_aug.num_lenders_total])}")
print(f"It is exactly a percentage of {len(loans_aug[loans_aug.lender != loans_aug.num_lenders_total])/len(loans_aug)}")

The number of different lenders number is 1121429
It is exactly a percentage of 0.8082790602467299
